R/json_nest.R

```diff
@@ -24,11 +24,44 @@ json_nest <- function(.data, ..., .names_sep = NULL) {
 json_nest.data.frame <- function(.data, ..., .names_sep = NULL) {
   check_suggested("jsonlite", use = TRUE, top_level_fun = "json_nest")
   dot_nms <- ...names()
-  # `{tidyr}` only warns but since we don't need backward compatibility we're
-  #   better off failing
-  if (is_null(dot_nms) || "" %in% dot_nms) {
-    abort("All elements of `...` must be named.")
-  }
   tidyr::nest(.data, ..., .names_sep = .names_sep) %>%
     mutate(across(all_of(dot_nms), ~ map_chr(., jsonlite::toJSON, digits = NA)))
 }
+
+#' @export
+json_nest.tbl_lazy <- function(.data, ..., .names_sep = NULL) {
+  dots <- quos(...)
+  if ("" %in% names2(dots)) {
+    abort("All elements of `...` must be named.")
+  }
+
+  col_nms <- colnames(.data)
+  nest_cols <- purrr::map(dots, ~ tidyselect::vars_select(col_nms, !!.x))
+  id_cols <- setdiff(col_nms, unlist(unique(nest_cols)))
+
+  sql_exprs <- purrr::imap(nest_cols, ~ sql_json_nest(
+    dbplyr::remote_con(.data),
+    cols = names(.x),
+    names_sep = .names_sep,
+    packed_col = .y
+  ))
+
+  .data %>%
+    group_by(across(all_of(id_cols))) %>%
+    summarize(!!!sql_exprs) %>%
+    ungroup()
+}
+
+sql_json_nest <- function(con, cols, names_sep, packed_col) {
+  UseMethod("sql_json_nest")
+}
+
+#' @export
+sql_json_nest.PqConnection <- function(con, cols, names_sep, packed_col) {
+  inside_cols <- remove_prefix_and_sep(cols, prefix = packed_col, sep = names_sep)
+  inside_cols_idented <- dbplyr::ident(inside_cols)
+  n <- length(inside_cols)
+  # alternate names and expressions for `json_build_object`
+  exprs <- c(syms(cols), inside_cols_idented)[rep(1:n, each = 2) + c(n, 0)]
```

Comment:
`1:n` is prone to corner cases, `seq_len(n)` is always safer.

I used `as.vector(t(matrix(..., ncol = 2)))` in similar cases, not sure which is better. With `!!!` we can ditch the `as.vector()`. Python has `zip()`, maybe we can add it as a helper function here?

Reply:
This is what I use normally, but in our case it's a bit more complex since we have a vector of characters and a list of symbols. We can adapt it though but it's less efficient.

``` r
x <- letters[1:3]
y <- LETTERS[1:3]
z <- rlang::syms(y)
as.vector(t(matrix(c(x,y), ncol = 2)))
#> [1] "a" "A" "b" "B" "c" "C"
as.vector(t(matrix(c(x,z), ncol = 2)))
#>      [,1] [,2] [,3]
#> [1,] "a"  "b"  "c" 
#> [2,] ?    ?    ?

unlist(t(matrix(c(x, purrr::map(z, list)), ncol = 2)))
#> [[1]]
#> [1] "a"
#> 
#> [[2]]
#> A
#> 
#> [[3]]
#> [1] "b"
#> 
#> [[4]]
#> B
#> 
#> [[5]]
#> [1] "c"
#> 
#> [[6]]
#> C
n <- length(x)
c(z, x)[rep(seq_len(n), each = 2) + c(n, 0)]
#> [[1]]
#> [1] "a"
#> 
#> [[2]]
#> A
#> 
#> [[3]]
#> [1] "b"
#> 
#> [[4]]
#> B
#> 
#> [[5]]
#> [1] "c"
#> 
#> [[6]]
#> C

X <- rep(x, 1e4)
Z <- rep(z, 1e4)
bench::mark(
  rep = c(Z, X)[rep(seq_len(n <- length(X)), each = 2) + c(n, 0)],
  matrix = unlist(t(matrix(c(X, purrr::map(Z, list)), ncol = 2)))
)
#> # A tibble: 2 × 6
#>   expression      min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 rep          1.21ms   1.33ms     739.     1.95MB     57.8
#> 2 matrix      10.22ms  10.28ms      96.6    2.06MB    869.
```

<sup>Created on 2022-04-28 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

Good idea to have a helper function
* `alternate <- function(...)` with `...` vectors/lists of same length ?
*  if we want it type stable in the `purrr` style we need `alternate()`, `alternate_chr()` etc
*  In this PR or a new one ?

Reply:
Let's keep this code as is, I opened a new issue for `alternate()` .

Reply:
https://github.com/cynkra/dm/pull/975/commits/4592c36123f95504b79e8951328e65165ee771dd and https://github.com/cynkra/dm/issues/982

Reply:
Sorry to interrupt but I just stumbled over your discussion about `alternate()`. I think that `vctrs::vec_interleave()` is what you are looking for

``` r
x <- letters[1:3]
y <- LETTERS[1:3]
vctrs::vec_interleave(x,y)
#> [1] "a" "A" "b" "B" "c" "C"

z <- rlang::syms(y)
vctrs::vec_interleave(as.list(x), z, .ptype = list())
#> [[1]]
#> [1] "a"
#> 
#> [[2]]
#> A
#> 
#> [[3]]
#> [1] "b"
#> 
#> [[4]]
#> B
#> 
#> [[5]]
#> [1] "c"
#> 
#> [[6]]
#> C
```

<sup>Created on 2022-04-29 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>